plugins/multifundchannel.c

```diff
@@ -0,0 +1,1939 @@
+
+static struct command_result *
+after_fundpsbt(struct command *cmd,
+	       const char *buf,
+	       const jsmntok_t *result,
+	       struct multifundchannel_command *mfc);
+
+static struct command_result *
+perform_fundpsbt(struct multifundchannel_command *mfc)
+{
+	struct out_req *req;
+
+	/* If the user specified utxos we should use utxopsbt instead
+	 * of fundpsbt.  */
+	if (mfc->utxos_str) {
+		plugin_log(mfc->cmd->plugin, LOG_DBG,
+			   "mfc %"PRIu64": utxopsbt.",
+			   mfc->id);
+
+		req = jsonrpc_request_start(mfc->cmd->plugin,
+					    mfc->cmd,
+					    "utxopsbt",
+					    &after_fundpsbt,
+					    &mfc_forward_error,
+					    mfc);
+		json_add_jsonstr(req->js, "utxos", mfc->utxos_str);
+		json_add_bool(req->js, "reservedok", false);
+	} else {
+		plugin_log(mfc->cmd->plugin, LOG_DBG,
+			   "mfc %"PRIu64": fundpsbt.",
+			   mfc->id);
+
+		req = jsonrpc_request_start(mfc->cmd->plugin,
+					    mfc->cmd,
+					    "fundpsbt",
+					    &after_fundpsbt,
+					    &mfc_forward_error,
+					    mfc);
+		json_add_u32(req->js, "minconf", mfc->minconf);
+	}
+
+	/* The entire point is to reserve the inputs.  */
+	json_add_bool(req->js, "reserve", true);
+	/* How much do we need to reserve?  */
+	if (mfc->has_all)
+		json_add_string(req->js, "satoshi", "all");
+	else {
+		struct amount_sat sum = AMOUNT_SAT(0);
+		for (size_t i = 0; i < tal_count(mfc->destinations); ++i) {
+			if (!amount_sat_add(&sum,
+					    sum, mfc->destinations[i].amount))
+				return mfc_fail(mfc, JSONRPC2_INVALID_PARAMS,
+						"Overflow while summing "
+						"destination values.");
+		}
+		json_add_string(req->js, "satoshi",
+				type_to_string(tmpctx, struct amount_sat,
+					       &sum));
+	}
+	json_add_string(req->js, "feerate",
+			mfc->feerate_str ? mfc->feerate_str : "normal");
+
+	{
+		size_t startweight;
+		size_t num_outs = tal_count(mfc->destinations);
+		/* Assume 1 input.
+		 * As long as lightningd does not select more than 252
+		 * inputs, that estimation should be correct.
+		 */
+		startweight = bitcoin_tx_core_weight(1, num_outs)
+			    + ( bitcoin_tx_output_weight( 1 /* OP_0 */
+							+ 1 /* OP_PUSHDATA */
+							+ 32 /* P2WSH */
+							)
+			      * num_outs
+			      )
+			    ;
+		json_add_string(req->js, "startweight",
+				tal_fmt(tmpctx, "%zu", startweight));
+	}
+
+	return send_outreq(mfc->cmd->plugin, req);
+}
+
+static struct command_result *
+compute_mfc_all(struct multifundchannel_command *mfc);
+static struct command_result *
+handle_mfc_change(struct multifundchannel_command *mfc);
+
+static struct command_result *
+after_fundpsbt(struct command *cmd,
+	       const char *buf,
+	       const jsmntok_t *result,
+	       struct multifundchannel_command *mfc)
+{
+	bool ok = true;
+	const jsmntok_t *field;
+
+	plugin_log(mfc->cmd->plugin, LOG_DBG,
+		   "mfc %"PRIu64": %s done.",
+		   mfc->id, mfc->utxos_str ? "utxopsbt" : "fundpsbt");
+
+	field = ok ? json_get_member(buf, result, "psbt") : NULL;
+	ok = ok && field;
```

Comment:
you can give better error messages if you stop at the problem and return a message there, rather than this "`ok` flag check for every field" strategy.

Reply:
Leads to much larger code that checks at each point (about 33% more, I tried with that already), and in practice giving the entire result when this errors helps debugging well enough.